src/raftstore/store/snap.rs

```diff
@@ -617,6 +618,750 @@ mod v1 {
+                    let mut value_bytes = kv.get_value();
+                    let size = box_try!(value_bytes.decode_u64());
+                    cf_sizes.push((cf, size));
+                }
+                SNAPSHOT_META_PREFIX_CHECKSUM => {
+                    let mut value_bytes = kv.get_value();
+                    let value = box_try!(value_bytes.decode_u64());
+                    if value > u32::max_value() as u64 {
+                        return Err(box_err!("invalid snapshot checksum {} for cf {}", value, cf));
+                    }
+                    cf_checksums.push((cf, value as u32));
+                }
+                _ => return Err(box_err!("invalid snapshot meta prefix {}", strs[0])),
+            }
+        }
+        if cf_sizes.len() != snapshot_cfs.len() {
+            return Err(box_err!("invalid number of snapshot meta cf size: {}, expected: {}",
+                                cf_sizes.len(),
+                                snapshot_cfs.len()));
+        }
+        if cf_checksums.len() != snapshot_cfs.len() {
+            return Err(box_err!("invalid number of snapshot meta cf checksum: {}, expected: {}",
+                                cf_checksums.len(),
+                                snapshot_cfs.len()));
+        }
+        let mut cf_size_checksums = Vec::with_capacity(snapshot_cfs.len());
+        for (cf, size) in cf_sizes {
+            let checksum = cf_checksums.iter().find(|c| c.0 == cf).unwrap().1;
+            cf_size_checksums.push((cf, size, checksum));
+        }
+        Ok(cf_size_checksums)
+    }
+
+    #[derive(Default)]
+    struct CfFile {
+        pub cf: String,
+        pub path: String,
+        pub size: u64,
+
+        // for building snapshot
+        pub sst_writer: Option<SstFileWriter>,
+        pub kv_count: u64,
+        // for sending/receiving snapshot
+        pub file: Option<File>,
+
+        // for writing snapshot
+        pub tmp_path: String,
+        pub written_size: u64,
+
+        pub digest: u32,
+    }
+
+    #[derive(Default)]
+    struct MetaFile {
+        pub data: RaftSnapshotData,
+        pub path: String,
+        pub file: Option<File>,
+
+        // for writing snapshot
+        pub tmp_path: String,
+    }
+
+    pub struct Snap {
+        display_path: String,
+        cf_files: Vec<CfFile>,
+        cf_index: usize,
+        meta_file: MetaFile,
+        size_track: Arc<RwLock<u64>>,
+    }
+
+    impl Snap {
+        fn new<T: Into<PathBuf>>(dir: T,
+                                 key: &SnapKey,
+                                 size_track: Arc<RwLock<u64>>)
+                                 -> io::Result<Snap> {
+            let dir_path = dir.into();
+            if !dir_path.exists() {
+                try!(fs::create_dir_all(dir_path.as_path()));
+            }
+            let prefix = format!("{}_{}", SNAP_GEN_PREFIX, key);
+            let display_path = Snap::get_display_path(&dir_path, &prefix);
```

Comment:
It's commonly seen that a single directory is used for raw data files, e.g. RocksDB use a single db dir to store sst files of all LSM levels.